Issue 609, 847: Add Terraform for automated service provisioning (#977)
raft-tech/TANF-app: terraform/README.md

## Prior Art

### Persistent vs Ephemeral Infrastructure
Adapted from: <https://github.com/HHS/Head-Start-TTADP#persistent-vs-ephemeral-infrastructure>

**The infrastructure used to run this application can be categorized into two distinct types: _ephemeral_ and _persistent_**

* **Ephemeral infrastructure** is all the infrastructure that is recreated each time the application deploys. Ephemeral infrastructure includes the "application(s)" (as defined in Cloud.gov), the EC2 instances the application runs on, and the routes that application utilizes. Our CircleCI configuration describes this infrastructure and deploys it to Cloud.gov.
* **Persistent infrastructure** is all the infrastructure that remains constant and unchanged despite application deployments. Persistent infrastructure includes the database used in each development environment. Our Terraform configuration files describe this infrastructure and instantiates it on Cloud.gov.

> This concept is often referred to as _mutable_ vs _immutable_ infrastructure.

### Infrastructure as Code

> A high-level configuration syntax, called [Terraform language][language], describes our infrastructure. This allows a blueprint of our system to be versioned and treated as we do any other code. This configuration can be acted on locally by a developer if deployments need to be created manually, but it is mostly and ideally executed by CirclCI.

Reviewer: Typo, CircleCI*

### Terraform state

We use an S3 bucket created by Cloud Foundry in Cloud.gov as our remote backend for Terraform. This backend maintains the "state" of Terraform and makes it possible for us to make automated deployments based on changes to the Terraform configuration files. **This is the only part of our infrastructure that must be manually configured.**

Note that a single S3 bucket maintains the Terraform State for both the development and staging environments, and this instance is deployed in the development space.

|   | development  | staging  | production | 
|---|---|---|---|
| S3 Key | `terraform.tfstate.dev`   | `terraform.tfstate.staging`  | `terraform.tfstate.prod`  |
| Service Space | `tanf-dev`  | `tanf-dev`  | `tanf-prod`  |

## Local Set Up For Manual Deployments

Sometimes a developer will need to run Terraform locally to perform manual operations. Perhaps a new TF State S3 bucket needs to be created in another environment, or there are new services or other major configuration changes that need to be tested first.

> 

Reviewer 2: @jorgegonzalez can we add the install steps for jq and circleci (in case anyone needs it)?

Author: Thanks, added!

1. **Install terraform**

    - On macOS: `brew install terraform`
    - On other platforms: [Download and install terraform][tf]

1. **Install Cloud Foundry CLI tool**

    - On macOS: `brew install cloudfoundry/tap/cf-cli`
    - On other platforms: [Download and install cf][cf-install]

1. **Login to Cloud Foundry**
    ```bash
       # login
       cf login -a api.fr.cloud.gov --sso
       # Follow temporary authorization code prompt.
       
       # Select the target org (probably `hhs-acf-prototyping`), 
       # and the space within which you want to provision infrastructure.
       
       # Spaces:
       # dev = tanf-dev
       # staging = tanf-staging
       # prod = tanf-prod
   ```

1. **Set up Terraform environment variables**

   In the `/terraform` directory, you can run the `create_tf_vars.sh` script which can be modified with details of your current environment, and will yield a `variables.tfvars` file which must be later passed in to Terraform. For more on this, check out [terraform variable definitions][tf-vars].

   ```bash
   ./create_tf_vars.sh
   

>    # Should generate a file `variables.tfvars` in the current directory.

Reviewer 3: ```suggestion
   # Should generate a file `variables.tfvars` in the `/terraform/dev` directory.
```

   # Your file should look something like this:
   #
   # cf_user = "some-dev-user"
   # cf_password = "some-dev-password"
   # cf_space_name = "tanf-dev"
   ```
   
## Test Deployment in Development

1. Follow the instructions above and ensure the `variables.tfvars` file has been generated with proper values.
1. `cd` into `/terraform/dev`

1. Prepare terraform backend:
   
   **Remote vs. Local Backend:**
   

>    If you merely wish to test some new changes without regards to the currently deployed remote state stored in the, you may want to use a "local" backend with Terraform.

Reviewer 2: is a word missing here?

Author: ```suggestion
   If you merely wish to test some new changes without regards to the currently deployed state stored in the remote TF state S3 bucket, you may want to use a "local" backend with Terraform.
```

Author: Yup, fixed!

   ```terraform
   terraform {
    backend "local" {}
   }
   ```
   

>    With this change, you should be able to run `terraform init` successfully.

Reviewer 2: change to `main.tf`?

Author: Correct, adding a clarification here.

Author: ```suggestion
   With this change to `main.tf`, you should be able to run `terraform init` successfully.
```
